### PR TITLE
Add an example of linking a rust_library into a cc_binary

### PR DESCRIPTION
**Pull request by bsilver8192**
allocator_library is now exposed at a higher level to support this.

This required swapping the order of libstd and allocator_library in the link. allocator_library is moving earlier, so this should enable it to depend on strictly more libraries than anything which worked before.

### CODE REVIEW

**Review by krasimirgg — APPROVED** [review]
This looks good. Thank you!

File: examples/cc_binary/forward_allocator.c
Comment (keith) [inline examples/cc_binary/forward_allocator.c:1]: should we move this file to a shared location where consumers of the rules can use it as well?

Comment (bsilver8192) [reply examples/cc_binary/forward_allocator.c:1]: Do you want to maintain it as a supported library here, or ask each project that uses it to maintain a version compatible with their rust toolchains? As you can see from the change I made yesterday, it does require updates for new Rust toolchains sometimes, because this is no stable interface to this part of the Rust ABI.

Comment (keith) [reply examples/cc_binary/forward_allocator.c:1]: Yea I can't say for sure since I'm not a maintainer on this, but as a user I would love it to be maintained here for everyone as opposed to me having to copy paste 🙃

Comment (krasimirgg) [reply examples/cc_binary/forward_allocator.c:1]: I think this is good as an example, but at least right now IMO it doesn't seem like something to add to a shared location and to commit rules_rust to maintain a specific version of this. My main concern is that this is subtle enough that in some non-trivial environments it would require some workspace-specific tweaks that will be hard to maintain and not be universal for all users. For example, we've got an environment where we have to weaken some of the `__rust*` symbols, otherwise we may run into duplicate symbol errors in some configurations when the dependency graph is sufficiently complicated.
I think it's worthwhile to follow-up the conversation about this over at https://github.com/bazelbuild/rules_rust/issues/1238.

@hlopko also may have thoughts about this.

**Comment by bsilver8192** [comment]
I'm happy to merge main to update this if the maintainers want to merge it, let me know. I definitely understand not wanting to maintain these hacks though.

**Comment by krasimirgg** [comment]
> I'm happy to merge main to update this if the maintainers want to merge it, let me know. I definitely understand not wanting to maintain these hacks though.

This PR looks good to me, ready to merge after updating on top of the main branch.

File: rust/toolchain.bzl
Comment (krasimirgg) [inline rust/toolchain.bzl:212]: Wouldn't it be more precise instead of switching the order here (effectively declaring allocator_inputs to depend on all of rust stdlib), to have allocator inputs depend just on alloc_inputs?

File: examples/cc_binary/forward_allocator.c
Comment (krasimirgg) [inline examples/cc_binary/forward_allocator.c:28]: could you add `extern "C"` before `__attribute__((weak))` of the __rust_* definitions? Without it, there is risk that these names get mangled in some configurations.

**Comment by bsilver8192** [comment]
#1490 addresses my use case perfectly. I don't think there's anything useful this PR adds on top of it.

**Pull request closed without merging**
